### ISSUE TITLE
Auto-tupling

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#66]
See https://github.com/samuelgruetter/srewrite-test/compare/diffdemo1original...diffdemo1#diff-9c0d11181aa5ea69a38f9de1ec4fffd6

**Comment by gabro** [comment]
I'm still wrapping my head around scalafix (as you may have noticed :) ). Any pointers on how to approach this? I'd like to give it a try!

**Comment by olafurpg** [comment]
I noticed :) 

The way I would approach this one would be to start writing a small unit test


```
<<< ONLY auto-tupling
object tup {
  def fooo(t: (Int, Int)): Int = ???
  fooo(1, 2)
}
>>>
object tup {
  def fooo(t: (Int, Int)): Int = ???
  fooo((1, 2))
}
```

and then create a semantic Rewrite

```
case class NoAutoTupling(mirror: Mirror) extends SemanticRewrite(mirror) {
  override def rewrite(ctx: RewriteCtx): Patch = {
    logger.elem(mirror.database)
    ???
  }
}
```

And then you should see something like

```
ExplicitArgs:11 [mirror.database]:
/var/folders/30/6jlz_xfs46ndvn212mt_wj6m0000gn/T/paradise3423812327028985194.scala
----------------------------------------------------------------------------------

Names:
[8..19): autotupling => _root_.autotupling.
[29..32): tup => _root_.autotupling.tup.
[41..45): fooo => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.
[46..47): t => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.(t)
[50..53): Int => _root_.scala.Int#
[55..58): Int => _root_.scala.Int#
[62..65): Int => _root_.scala.Int#
[68..71): ??? => _root_.scala.Predef.`???`()Lscala/Nothing;.
[74..78): fooo => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.
```

The interesting bit is the last line which shows that `fooo` takes a single argument of Tuple2. That should be a good indicator that auto-tupling is happening. You can query for the symbol with the extension method `Ref.symbol` as long as an implicit Mirror is in scope. `Ref` is a supertype of `Term.Name` like `foo` in `foo(1, 2)`.


FWIW . I have a local branch (https://github.com/olafurpg/scalafix/compare/meta-1.8?expand=1) that uses scalameta from https://github.com/scalameta/scalameta/pull/808 which adds two more sections to `mirror.database`

```
ExplicitArgs:11 [mirror.database]:
/var/folders/30/6jlz_xfs46ndvn212mt_wj6m0000gn/T/paradise5845667176885744242.scala
----------------------------------------------------------------------------------

Names:
[8..19): autotupling => _root_.autotupling.
[29..32): tup => _root_.autotupling.tup.
[41..45): fooo => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.
[46..47): t => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.(t)
[50..53): Int => _root_.scala.Int#
[55..58): Int => _root_.scala.Int#
[62..65): Int => _root_.scala.Int#
[68..71): ??? => _root_.scala.Predef.`???`()Lscala/Nothing;.
[74..78): fooo => _root_.autotupling.tup.fooo(Lscala/Tuple2;)I.

Denotations:
_root_.autotupling. => package autotupling
_root_.autotupling.tup. => final object tup
_root_.autotupling.tup.fooo(Lscala/Tuple2;)I. => def fooo: (t: (Int, Int))Int
_root_.autotupling.tup.fooo(Lscala/Tuple2;)I.(t) => param t: (Int, Int)

Sugars:
[84..84) [Int, Int]
```
However, I think symbols are enough to detect when auto-tupling is happening.

**Comment by gabro** [comment]
Closed by #143 and superseded by #147? @olafurpg

**Comment by olafurpg** [comment]
Closed and superseded by #147 Thank you @gabro for contributing this rewrite!
